### ISSUE TITLE
sysupdate should allow for upgrading to a newer major version

### ISSUE THREAD

**Issue by abderrahim** [systemd/systemd#33345]
### Is your feature request related to a problem? Please describe

Say I'm on foobarOS 46, and a new major version 47 is released. I want to be able to keep getting updates for foobarOS 46 automatically. I also want to be able to easily upgrade to foobarOS 47.

### Describe the solution you'd like

The proposed approach revolves around 
* Use both `VERSION_ID` and `IMAGE_VERSION` (from os-release) to identify the image version, where `VERSION_ID` is the major version.
* Only upgrade to a version with a `VERSION_ID` matching the current install by default
* Allow overriding the `VERSION_ID` to perform a major version upgrade

I still need to wrap my head around match patterns and specifiers to see if we can do this without changing the current semantics, but my current idea is:
* change the os-release spec to say: VERSION_ID should/must be prefix of IMAGE_VERSION
* when filtering candidates, sysupdate should only consider versions where the current VERSION_ID matches a prefix of IMAGE_VERSION
* sysupdate/updatectl get a way to override the VERSION_ID to make the update target a newer major version.

### The systemd version you checked that didn't have the feature you are asking for

256

**Comment by AdrianVovk** [comment]
I think there's a much simpler thing we can do on the systemd side, and leave the complexity (but also flexibility) up to the distributor.

My idea was to add overrides for the parsing of the specifiers themselves. So let's say you have a specifier %F that resolves to the env var FOOBAR in /etc/os-release. We could teach the systemd-sysupdate tool a `--override=FOOBAR=someothervalue`. Then whenever the %F specifier appears, it would be substituted with `someothervalue` instead of whatever appears in /etc/os-release

The docs already state that VERSION_ID is for major version numbers of the OS. We should make those docs more clear. But for an image-based OS, the solution would look something like this. Make your update URL (I'll explicitly write out the os-release names instead of looking up the specifiers right now): `https://dl.example.com/updates/v1/$IMAGE_ID/$VARIANT_ID/$VERSION_ID/<file>.$IMAGE_VERSION.<etc>`. So the usr image of the generic x86 variant of exampleOS 46.3 would be `https://dl.example.com/updates/v1/generic-x64/desktop/46/usr.46.3.<etc>`. Then if `--override=VERSION_ID=47` is passed in, and we're looking for the latest $IMAGE_VERSION (i.e. the `@v` pattern) then it'll look for the latest usr tree under `.../generic-x64/desktop/47/SHA256SUMS`.

We can wire this `--override` up to the sysupdated API trivially. I would call the function something like `Switch` (similar to how you'd `ostree admin switch` to a different branch). Then leave the implementation of what calls `Switch` up to the distributor. Could be a special update center like Ubuntu or Linux Mint, could be built into the normal updater like Fedora, could be a special command, whatever

**Comment by abderrahim** [comment]
> The docs already state that VERSION_ID is for major version numbers of the OS. We should make those docs more clear. But for an image-based OS, the solution would look something like this. Make your update URL (I'll explicitly write out the os-release names instead of looking up the specifiers right now): `https://dl.example.com/updates/v1/$IMAGE_ID/$VARIANT_ID/$VERSION_ID/<file>.$IMAGE_VERSION.<etc>`. So the usr image of the generic x86 variant of exampleOS 46.3 would be `https://dl.example.com/updates/v1/generic-x64/desktop/46/usr.46.3.<etc>`. Then if `--override=VERSION_ID=47` is passed in, and we're looking for the latest $IMAGE_VERSION (i.e. the `@v` pattern) then it'll look for the latest usr tree under `.../generic-x64/desktop/47/SHA256SUMS`.

The problem I see with this approach is that systemd-sysupdate would have no way of knowing that version 47 exists, unless I'm passing that override. So I have to get that info from a separate channel.

**Comment by AdrianVovk** [comment]
Yes this is something I was trying to solve with the appstream metadata.

I wonder, maybe we can have a special entry added to SHA256SUMS that offers overrides to upgrade to? Or just a file next to SHA256SUMS? So we're on v46, and now v47 is out, so we make an OVERRIDES file next to SHA256SUMS that says `VERSION_ID=47`. Then sysupdate will know to offer it

**Comment by tomcoldrick-ct** [comment]
I wonder if it might be better to do this outside of sysupdate altogether.  Each OS version could be a "component" with a `/etc/sysupdate.myOS-vX.d`, with the active one defined by a symlink to the host component `/etc/sysupdate.d`.  This would allow us to get updates for all such versions, and have a single one active.  The implementation of the `switch` verb can be controlling this symlink.

My reasoning here is that overrides only work if the set of transfers is static between major releases.  If a new release has a new structure, then an override may well just break on download.  In the standard case of a single set of OS images, (UKI plus a usr partition, say) then this is may well be unlikely, but we're also hoping to use sysupdate to define, say sysexts.  If we split or  rename a sysext between major versions, then the transfer will fail, and we won't update.  I think it makes more sense for the OS image _itself_ to contain the sysupdate configuration for the next release, and let the user enable it when they're ready.

The flow would be something like an update to the current image including an optional (but disabled) configuration.  Users can `updatectl enable` this to download the system, and `updatectl switch` to switch to the new major release.

**Comment by AdrianVovk** [comment]
The --override mechanism is useful for more than just major OS updates. It's useful for switching between board IDs. Imagine you've installed an x86_64_v1 baseline edition of the OS, but in the future we start releasing an x86_64_v3 build. You would want to switch your board from `x86_64-generic` to `x86_64_v3-generic` if your CPU supports x86_64_v3, or stay on x86_64 if it doesn't.

Or let's say the designers really want to implement [these mockups](https://gitlab.gnome.org/Teams/Design/os-mockups/-/blob/master/mobile-shell/mobile-shell-convergence.png) but want feedback first, so we start releasing a new `convergence` variant of the OS as a long-running experiment to see what user response would be. Users should be able to opt into this experiment, and then opt back out, or once the experiment ends let it seamlessly roll back into the stable release.

All that said, this might not necessarily need to be part of sysupdated directly, nor does it need to be involved in the major system upgrade process...  We could just as easily have a script `gnomeos-switch-board <newboard>` that stops `systemd-sysupdated.service`, then masks it in /run, then copies over `/usr/llib/sysupdate.d` into some temporary location, then `sed`s all the specifiers to be overriden, then runs `systemd-sysupdate --definitions ...`, then finally reboots the system. We can do similar things with variants. For these kinds of manual transitions, where the user is opting into them, I think that's good enough. So let's move past `--override`.

---

I like where you're going with your proposal w/ a component, but we probably can simplify it. It also has some bugs: if we define a component, `updatectl update` will always install it and break the system. So here's what I'm counter-proposing:

We define `/usr/lib/sysupdate-next.d` and `/usr/lib/sysregistry-next.d`. These correspond to the `/usr/lib/sysupdate.d` and `/usr/lib/sysregistry.d` of the next major version of the host OS. You can access them via `/usr/lib/systemd/systemd-sysupdate --next` (so that way we have all the fallback locations into /etc and /run if the user really wants to).

When we're building GNOME OS, we separately package and install the sysupdate config and syspdate-next config (if the latter is available). We will have a policy that states: changes to the sysupdate config must happen relatively early in the development cycle of the next major release (unless it's absolutely necessary), and the moment we deem it stabilized we immediately backport it from nightly into the current stable's sysupdate-next config. This gives time for this next config to propagate to users via regular security patches to the current stable branch, and avoids the situation where the user must install the latest patch on the Nth stable branch only to be immediately offered an update to the (N+1)th stable branch. In other words, it should be possible to update from a relatively early patch of the Nth release directly to any patch of the (N+1)th release.

<details>
<summary>More suggestions how to implement this in gnome-build-meta</summary>

Whenever it's time to cut a new stable release, we just move the sysupdate-next config over top of the old sysupdate config. Personally, the way I would do it over on the buildstream side is have something like `files/sysupdate-config` and `files/sysupdate-next-config`. When packaging, we'd just verbatim copy `files/sysupdate-config` into `/usr/lib/sysupdate.d`, but for `files/sysupdate-next-config` we'd run `sed` over it to replace all instances of `%w` with the next major version number before copying it into `/usr/lib/sysupdate-next.d`. Ditto with sysregistry. The exact layout of filenames doesn't matter here, the point is the `sed` substitution while leaving the `%w` variable in the source files. That way, when it's time to cut a new release, we can quite literally just move `files/sysupdate-next-config` over top of `files/sysupdate-config` and be done with it, without needing to touch the packaging or the configuration inside the files. And when backporting the config files from nightly, we'd just backport them 100% unchanged also.

</details>

sysupdated would treat this as a new class, called `host-next`. It would behave exactly like `host`, but it would pass `--next` into systemd-sysupdate (which of course would tell it to use the `sysupdate-next.d` directories). Until the next major release is actually released, there will be no available versions: the current version will be detected but will be unavailable (so even if it's seen as an incomplete version sysupdate will not offer to complete it). Once the next major release is out, however, sysupdate will be able to offer it as a normal update to the `host-next` target. Of course, installing that update would install into the same set of partitions as the current `host` target is using, exactly like we want - it'll wipe out the current oldest host slot, applying all version protections and whatnot (thanks to my work in https://github.com/systemd/systemd/pull/33398 the protections will work even if the current host is seen as incomplete because host-next has new transfers), and install into that space the next major release. This `host-next` target is only offered if there are `sysupdate-next.d` config files; if not, it's not offered by sysupdated.

NOTE: sysupdated would need a special case, that whenever the `host-next` is being manipulated the `host` target is also marked as busy, and vice versa. This avoids situations where a user tries to upgrade both `host` and `host-next` at the exact same time and ends up with two instances of systemd-sysupdate talking over each other

`updatectl` needs to special-case this `host-next` target to make sure it is never automatically selected for anything destructive. That way, `updatectl update` doesn't accidentally bump your major OS version. The user would have to specify `updatectl update host-next` to bump to the next major version. We would automatically check `host-next` for updates, and list versions, hiding any version that's not marked `AVAILABLE` (since non-AVAILABLE versions in `host-next` are actually the currently installed versions in `host`) (maybe we apply this filtering over on the sysupdated side so that clients only ever see AVAILABLE versions of `host-next`? I'm not sure... But the sysupdate worker process must not special case this or else things break). We must prevent the user from operating on `host` and `host-next` at the same time in all cases. These tweaks should make the experience seamless on the command-line.

`gnome-software` would check the `host-next` target for updates just like it checks `host`. It's a separate target that is updated independently, as far as Software is concerned. Whenever `host-next` claims to have an update available, it's presented with the major update banner. Note that it's perfectly valid for both `host` and `host-next` to offer available updates! The user can update `host` to upgrade the patch level of the current release, or update `host-next` to switch to the next major version. If the user starts upgrading one, Software should prevent the user from trying to upgrade the other.

---

Phew, that was dense. Sorry about that, I want to be complete in my detailing of the implementation, so please refer there for all details. But let's summarize the important points:

- We define new `sysupdate-next.d` and `sysregistry-next.d` directories. These are, essentially, just the `sysupdate.d` and `sysregistry.d` directories of the next major release.
- sysupdated exposes these directories in a new target called `host-next`. Updating `host-next` upgrades you to the next major release. We've got two small special cases (sharing busy flag, and filtering out non-available versions) to implement in sysupdated. Everything else basically just happens to work as intended,
- We have policies over on the distro development side to ensure that running a relatively up-to-date build of the current stable release will allow you to upgrade directly to any patch level of the next major release.
- `updatectl` will need special treatment of the `host-next` target to avoid accidentally damaging the host OS or upgrading without explicit user input. The user would do `updatectl update host-next` to upgrade to the latest patch of the next major release (or specify an exact version like usual if they want a specific patch of the next major release).
- `gnome-software` would display available updates to `host-next` using the special major upgrade UI. It will also need to prevent the user from trying to update both `host` and `host-next` at the same time, but it should still offer both updates until the user picks one.

**Comment by chunyi-wu** [comment]
In this `sys{update,update-next}.d` approach, what happens to the components? Do we need to support `sys{update-next.<component>}.d` as well, or we assume they are unrelated to the OS image?

**Comment by AdrianVovk** [comment]
sysupdate components are, by definition, updated independently from the OS version. So they aren't relevant. Ditto with portable services, machined machines, etc.

As for extensions, we'd have a `sysregistry-next.d` too, containing definitions for the next version's optional transfers. I'll rework #33398 to better support this situation, and give more detail on exactly what you need to do to make it work with `--next` once I nail down all the details.

**Comment by AdrianVovk** [comment]
OK, I've put together a more concrete idea in my head what the replacement for `sysregistry` might look like, and put that idea into words here: https://github.com/systemd/systemd/issues/33343#issuecomment-2192402869.

For major version updates to work, we'd need to apply `/etc/sysfeatures.d/*.feature.d/enable.conf` onto the features defined in `/usr/lib/sysfeatures.d`. More concretely, the system administrator may be overriding any number of things in `/etc` or `/run` that might apply to `sysfeatures-next.d`. And when the upgrade goes through, those files in /etc _will_ be applying to the next `/usr/lib/sys{update,features}.d`.

So my previous proposal was that the serach path for `--next` should look like: `/usr/lib/sys{update,features}-next.d`, `/etc/sys{update,features}-next.d`, `/run/sys{update,features}-next.d`. Maybe instead it should look more like this: `/usr/lib/sys{update,features}-next.d`, `/etc/sys{update,features}.d`, `/run/sys{update,features}.d`. Or like this:
`/usr/lib/sys{update,features}-next.d`, `/etc/sys{update,features}-next.d`, `/run/sys{update,features}-next.d`,  `/etc/sys{update,features}.d`, `/run/sys{update,features}.d`. Open to ideas here!

Alternatively, we could just loop over all the available features in `sysfeatures.d/`, mark down which are enabled, then look up the features defined in `sysfeatures-next.d/` and treat any feature with a matching name as enabled. But then I'm not sure how we'd handle situations where the drop-ins in /etc are overriding something else about the feature, or the transfers for that matter. Should we merge those in? IDK

**Comment by chunyi-wu** [comment]
Assumes that I notice there is one new feature in the 'next' version I would like use.

I enable in to the `/etc/sysfeatures- next.d/<new-feature>.feature.d/enable.conf`, then upgrade with `--next`.
After upgrade to the next version and reboot, that feature may not be enabled since the `enable.conf ` still lay in `/etc/sysfeatures-next.d` instead of `/etc/sysfeatures.d` I think. Also, I looks kind of weird to me to put the enable under the `/etc/sysfeatures.d` before actually upgrade to the target version.

As a result, I think unless we are moving configuration files under `/etc`, the directory `/etc/sysfeatures-next.d` looks weird to me.

**Comment by AdrianVovk** [comment]
> Assumes that I notice there is one new feature in the 'next' version I would like use.

In that situation you really should probably upgrade the OS first, then enable the feature... We will have no GUI or CLI support for both enabling a feature and upgrading to the next major OS version at the same time.

**Comment by chunyi-wu** [comment]
> > Assumes that I notice there is one new feature in the 'next' version I would like use.
> 
> In that situation you really should probably upgrade the OS first, then enable the feature... We will have no GUI or CLI support for both enabling a feature and upgrading to the next major OS version at the same time.

I fully agree with you. What I am thinking is in what cases do we need to put configurations under `{run,etc}/sys{update,features}-next.d`?

**Comment by AdrianVovk** [comment]
my concern is that having certain overrides from `/etc/sysfeatures.d/` apply onto `/usr/lib/sysfeatures-next.d` is 1) unprecedented and thus unexpected, 2) may cause the changes in /etc to apply onto completely different configurations than they intended to.

On the other hand, the `-next.d` directories become the normal `.d` directories once the update is complete, so perhaps this isn't a concern really. I just want input about how others feel about it. (i.e. @poettering)
